state/ravendb/ravendb.go

```diff
@@ -0,0 +1,520 @@
+	metadataStruct := RavenDBMetadata{}
+	metadata.GetMetadataInfoFromStructType(reflect.TypeOf(metadataStruct), &metadataInfo, metadata.StateStoreType)
+	return
+}
+
+func (r *RavenDB) setInternal(ctx context.Context, req *state.SetRequest, session *ravendb.DocumentSession) error {
+	data, err := r.marshalToString(req.Value)
+	if err != nil {
+		return fmt.Errorf("ravendb error: failed to marshal value for key %s: %w", req.Key, err)
+	}
+
+	item := &Item{
+		ID:    req.Key,
+		Value: data,
+	}
+
+	if req.Options.Concurrency == state.FirstWrite {
+		// First write wins, we send empty change vector to check if exists
+
+		// current SDK version of go doesn't let us to check concurency violation on items that are not in databse.
+		// we need to try to load, and do regullar save if item is not in DB (real first save)
+		// if we have item in DB we can try to override it with concurency check
+		var newItem *Item
+		err = session.Load(&newItem, req.Key)
+		if newItem == nil {
+			err = session.Store(item)
+		} else {
+			var eTag string
+			if req.HasETag() {
+				eTag = *req.ETag
+			} else {
+				eTag = RandStringRunes(5)
+			}
+
+			if newItem.Value == item.Value {
+				return fmt.Errorf("error storing data: %s", err)
+			}
+			newItem.Value = item.Value
+			err = session.StoreWithChangeVectorAndID(newItem, eTag, req.Key)
+		}
+		if err != nil {
+			return fmt.Errorf("error storing data: %s", err)
+		}
+	} else {
+		// Last write wins
+		if req.HasETag() {
+			eTag := *req.ETag
+			err = session.StoreWithChangeVectorAndID(item, eTag, req.Key)
+			if err != nil {
+				return state.NewETagError(state.ETagMismatch, err)
+			}
+		} else {
+			err = session.Store(item)
+		}
+
+		if err != nil {
+			return fmt.Errorf("error storing data: %s", err)
+		}
+	}
+
+	reqTTL, err := stateutils.ParseTTL(req.Metadata)
+	if err != nil {
+		return fmt.Errorf("failed to parse TTL: %w", err)
+	}
+
+	if reqTTL != nil {
+		metaData, err := session.Advanced().GetMetadataFor(item)
+		if err != nil {
+			return errors.New("failed to get metadata for item")
+		}
+		expiry := time.Now().Add(time.Second * time.Duration(*reqTTL)).UTC()
+		iso8601String := expiry.Format("2006-01-02T15:04:05.9999999Z07:00")
+		metaData.Put(expires, iso8601String)
+	}
+	return nil
+}
+
+func (r *RavenDB) deleteInternal(ctx context.Context, req *state.DeleteRequest, session *ravendb.DocumentSession, fromTransaction bool) error {
+	var err error
+	if fromTransaction {
+		var itemToDelete *Item
+		err = session.Load(&itemToDelete, req.Key)
+		if err == nil {
+			err = session.Delete(itemToDelete)
+		}
+	} else {
+		if req.HasETag() {
+			err = session.DeleteByID(req.Key, *req.ETag)
+		} else {
+			// TODO: Fix after update to ravendb sdk
+			err = session.DeleteByID(req.Key, "")
+		}
+	}
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *RavenDB) getRavenDBStore(ctx context.Context) (*ravendb.DocumentStore, error) {
```

Comment:
This could also be a `setRavenDBStore` and set `r.documentStore` directly, instead of returning the store object.